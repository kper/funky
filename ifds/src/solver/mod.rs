//! The implementation for extracting the taints from the graph.

use crate::icfg::graph::{Edge, Graph};
use anyhow::{Context, Result};
use std::collections::HashSet;

type PC = usize;

pub struct IfdsSolver;

/// Represents the taints from the execution of the solver.
#[derive(Debug)]
pub struct Taint {
    pub variable: String,
    pub function: String,
    pub pc: PC,
}

/// Request is the datastructure which defines at what instruction the [`Graph`] should be build from.
#[derive(Debug, Clone)]
pub struct Request {
    pub variable: Option<String>,
    pub function: String,
    pub pc: PC,
}

pub trait Solver {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>>;

    /// Check if the statement at `req` and the statement `resp` have a taint.
    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool>;

    /// Return all tainted variables names at any statement for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>>;
}

impl Solver for IfdsSolver {
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints: Vec<_> = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    //&& x.get_from().var_is_taut
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| Taint {
                function: x.function.clone(),
                pc: x.next_pc.checked_sub(1).unwrap_or(0),
                variable: x.belongs_to_var.clone(),
            })
            .collect();

        Ok(taints)
    }

    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| x.belongs_to_var.clone())
            .collect::<HashSet<_>>();

        Ok(taints)
    }

    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool> {
        let all_sinks = self
            .all_sinks(graph, req)
            .context("Cannot compute the sinks")?;

        let var = resp
            .variable
            .as_ref()
            .context("Please specify a variable for the response")?;

        Ok(all_sinks
            .into_iter()
            .find(|x| x.function == resp.function && x.pc == resp.pc && &x.variable == var)
            .is_some())
    }
}

pub trait GraphReachability {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Vec<Taint>;
}
