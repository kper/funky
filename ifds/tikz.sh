#!/bin/bash

rm /tmp/graph.tex
rm /tmp/graph.pdf
tail -n +6 $1 > /tmp/graph.tex
xelatex /tmp/graph.tex && evince graph.pdf
