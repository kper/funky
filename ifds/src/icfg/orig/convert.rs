/// This module is responsible to parse
/// the webassembly AST to a graph
use crate::icfg::graph::*;
use crate::icfg::state::State;
use crate::ir::ast::Function as AstFunction;
use crate::ir::ast::Instruction;

use crate::solver::Request;
use anyhow::{Context, Result};
use std::collections::VecDeque;

use log::debug;

use crate::ir::ast::Program;

use crate::icfg::naive::convert::Convert as NaiveConvert;
use std::collections::HashMap;

type CallerFunction = String;
type PC = usize;
type Function = String;
type CallResolver = HashMap<Function, Vec<(CallerFunction, PC, Vec<String>)>>;
pub(crate) struct Ctx<'a> {
    pub graph: &'a mut Graph,
    pub new_graph: &'a mut Graph,
    pub state: &'a mut State,
    pub call_resolver: &'a CallResolver,
}

/// Central datastructure for the computation of the IFDS problem.
#[derive(Debug, Default)]
pub struct OriginalConvert;

impl OriginalConvert {
    /// Computes a graph by a given program and `req` ([`Request`]).
    /// The `variable` in `req` doesn't matter. It only matters the `function` and `pc`.
    pub fn visit(&mut self, prog: &Program, req: &Request) -> Result<(Graph, State)> {
        let mut fact_gen = NaiveConvert::default();
        let (mut graph, mut state, call_resolver) = fact_gen
            .visit(prog)
            .context("Naive fact generation failed")?;

        let mut new_graph = Graph::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            new_graph: &mut new_graph,
            call_resolver: &call_resolver,
        };

        self.tabulate(&mut ctx, prog, req)?;

        Ok((new_graph, state))
    }

    fn tabulate<'a>(&mut self, ctx: &mut Ctx<'a>, prog: &Program, req: &Request) -> Result<()> {
        debug!("Convert intermediate repr to graph");

        let function = prog
            .functions
            .iter()
            .find(|x| x.name == req.function)
            .context("Cannot find function")?;

        let facts = ctx
            .state
            .get_facts_at(&function.name, req.pc)?
            .cloned()
            .collect::<Vec<_>>();

        let mut path_edge = Vec::new();
        let mut worklist = VecDeque::new();
        let mut summary_edge = Vec::new();
        let mut normal_flows_debug = Vec::new();

        let init = facts.get(0).unwrap().clone();

        // self loop for taut
        self.propagate(
            &mut ctx.new_graph,
            &mut path_edge,
            &mut worklist,
            Edge::Path {
                from: init.clone(),
                to: init.clone(),
            },
        )?;

        self.forward(
            &prog,
            &function,
            &mut path_edge,
            &mut worklist,
            &mut summary_edge,
            &mut normal_flows_debug,
            ctx,
            req.pc,
        )?;

        Ok(())
    }

    fn propagate(
        &self,
        graph: &mut Graph,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        e: Edge,
    ) -> Result<()> {
        let from = e.get_from();
        let to = e.to();

        let f = path_edge
            .iter()
            .find(|x| x.get_from() == from && x.to() == to);

        if f.is_none() {
            debug!("Propagate {:#?}", e);
            graph.edges.push(e.clone());
            path_edge.push(e.clone());
            worklist.push_back(e);
        }

        Ok(())
    }

    fn forward<'a>(
        &mut self,
        program: &Program,
        _function: &AstFunction,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        summary: &mut Vec<Edge>,
        normal_flows_debug: &mut Vec<Edge>,
        ctx: &mut Ctx<'a>,
        start_pc: usize,
    ) -> Result<()> {
        while let Some(edge) = worklist.pop_front() {
            debug!("Popping edge from worklist {:#?}", edge);

            assert!(
                matches!(edge, Edge::Path { .. }),
                "Edge in the worklist has wrong type"
            );

            let d1 = edge.get_from();
            let d2 = edge.to();

            let pc = edge.to().next_pc;
            debug!("Instruction pointer is {}", pc);

            let instructions = &program
                .functions
                .iter()
                .find(|x| x.name == d2.function)
                .context("Cannot find function")?
                .instructions;
            let n = instructions.get(pc);
            debug!("=> Instruction {:?}", n);

            if let Some(n) = n {
                match n {
                    Instruction::Call(callee, _params, _dest) => {
                        // handle_call

                        let caller_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d1.function)
                            .unwrap();

                        let callee_function = program
                            .functions
                            .iter()
                            .find(|x| &x.name == callee)
                            .unwrap();

                        let flow_edges = ctx.graph.edges.iter().filter(|x| {
                            x.get_from().function == caller_function.name
                                && x.get_from().belongs_to_var == d2.belongs_to_var
                                && x.get_from().next_pc == d2.next_pc
                                && x.to().function == callee_function.name
                                && x.to().next_pc == 0
                        });

                        for f in flow_edges.into_iter() {
                            let to = f.to();
                            self.propagate(
                                &mut ctx.new_graph,
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: to.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }

                        let flow_edges = ctx
                            .graph
                            .edges
                            .iter()
                            .chain(summary.iter())
                            .filter(|x| {
                                x.get_from().function == caller_function.name
                                    && x.get_from().belongs_to_var == d2.belongs_to_var
                                    && x.get_from().next_pc == d2.next_pc
                                    && x.to().function == caller_function.name
                                    //&& x.to().belongs_to_var == x.get_from().belongs_to_var
                                    && x.to().next_pc == x.get_from().next_pc + 1
                            })
                            .map(|x| x.to());

                        for to in flow_edges.into_iter() {
                            assert_eq!(d1.function, to.function);

                            self.propagate(
                                &mut ctx.new_graph,
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d1.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }
                    }
                    _ => {
                        let new_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d2.function)
                            .unwrap();

                        let flow_edges =
                            ctx.graph
                                .edges
                                .iter()
                                .filter(|x| x.is_normal())
                                .filter(|x| {
                                    x.get_from().function == new_function.name
                                        && x.get_from().belongs_to_var == d2.belongs_to_var
                                        && x.get_from().next_pc == d2.next_pc
                                });

                        for f in flow_edges.into_iter() {
                            debug!("Normal flow {:#?}", f);
                            let to = f.to();

                            normal_flows_debug.push(f.clone());

                            self.propagate(
                                &mut ctx.new_graph,
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d1.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }
                    }
                }
            } else {
                // end procedure

                if let Some(callers) = ctx.call_resolver.get(&d1.function) {
                    let callers = callers
                        .iter()
                        .map(|(caller, _pc, _dests)| (caller, &d1.function));

                    for (caller, callee) in callers {
                        assert_eq!(callee, &d1.function);

                        let d4 = ctx
                            .graph
                            .edges
                            .iter()
                            .filter(|x| x.is_call())
                            .filter(|x| {
                                &x.get_from().function == caller
                                    && x.to().function == d1.function
                                    && x.to().belongs_to_var == d1.belongs_to_var
                                    && x.to().next_pc == d1.next_pc
                            })
                            .map(|x| x.get_from());

                        for d4 in d4.into_iter() {
                            let d5 = ctx
                                .graph
                                .edges
                                .iter()
                                .filter(|x| x.is_return())
                                .filter(|x| {
                                    x.get_from().belongs_to_var == d2.belongs_to_var
                                    && x.get_from().function == d2.function
                                    && x.get_from().next_pc == d2.next_pc - 1 //because applied before
                                    && &x.to().function == &d4.function
                                    // && x.to().next_pc == pc + 1
                                })
                                .map(|x| x.to());

                            for d5 in d5 {
                                if summary
                                    .iter()
                                    .find(|x| {
                                        &x.get_from().function == caller
                                            && x.get_from().belongs_to_var == d4.belongs_to_var
                                            && &x.to().function == caller
                                            && x.to().next_pc == x.get_from().next_pc + 1
                                            && x.to().belongs_to_var == d5.belongs_to_var
                                    })
                                    .is_none()
                                {
                                    let ret = d4.clone();
                                    summary.push(Edge::Summary {
                                        from: d4.clone(),
                                        to: ret.clone(),
                                    });

                                    let edges = ctx
                                        .graph
                                        .edges
                                        .iter()
                                        .filter(|x| {
                                            &x.get_from().function == caller
                                                && &x.to().function == caller
                                                && x.get_from().next_pc == start_pc
                                                && x.to().belongs_to_var == d4.belongs_to_var
                                                && x.to().next_pc == d4.next_pc
                                        })
                                        .map(|x| x.get_from());

                                    for d3 in edges {
                                        let mut ret = ret.clone();
                                        ret.belongs_to_var = d5.belongs_to_var.clone();

                                        let init = ctx
                                            .state
                                            .get_facts_at(caller, 0)?
                                            .find(|x| x.belongs_to_var == d3.belongs_to_var)
                                            .context("Cannot find fact")?
                                            .clone();

                                        self.propagate(
                                            &mut ctx.new_graph,
                                            path_edge,
                                            worklist,
                                            Edge::Path {
                                                from: init,
                                                to: ret,
                                            },
                                        )?;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        //graph.edges.extend_from_slice(&path_edge);
        //graph.edges.extend_from_slice(&normal_flows_debug);
        //graph.edges.extend_from_slice(&summary_edge);

        Ok(())
    }
}
