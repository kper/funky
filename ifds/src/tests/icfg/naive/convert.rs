use crate::icfg::naive::convert::Convert;
use crate::icfg::tikz::render_to;
use crate::solver::Request;
use insta::assert_snapshot;
use log::error;

use crate::grammar::*;

macro_rules! ir {
    ($name:expr, $req:expr, $ir:expr) => {
        let mut convert = Convert::default();

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let res = convert.visit(&prog);

        if let Err(err) = res {
            error!("ERROR: {}", err);
            err.chain()
                .skip(1)
                .for_each(|cause| error!("because: {}", cause));
            panic!("")
        }

        //write_ir($name, $ir);

        let (graph, state) = res.unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("{}_dot", $name), output);
    };
}

#[test]
fn test_ir_const() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_const",
        req,
        "
         define test (result 0) (define %0) {
            %0 = 1
         };
    "
    );
}


#[test]
fn test_ir_double_const() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_double_const",
        req,
        "
         define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = 1
         };
    "
    );
}

#[test]
fn test_ir_double_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_double_assign",
        req,
        "
         define test (result 0) (define %0 %1 %2){
            %0 = 1
            %1 = 1
            %2 = %0
            %2 = %1
         };
    "
    );
}

#[test]
fn test_ir_double_assign_with_params() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_double_const_with_params",
        req,
        "
         define test (param %0) (result 0) (define %0 %1 %2){
            %1 = 1
            %2 = %0
            %2 = %1
         };
    "
    );
}

#[test]
fn test_ir_chain_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_chain_assign",
        req,
        "
         define test (result 0) (define %0 %1 %2 %3){
            %0 = 1
            %1 = 1
            %2 = %0
            %3 = %2
         };
    "
    );
}

#[test]
fn test_ir_unop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_unop",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = op %0
            %1 = op %0   
        };"
    );
}

#[test]
fn test_ir_binop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_binop",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_binop_offset() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_binop_offset",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_phi() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_phi",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = phi %0 %1
        };"
    );
}

#[test]
fn test_ir_killing_op() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_killing_op",
        req,
        "define test (result 0) (define %0 %1 %2)  {
            %0 = 1
            %1 = 1
            KILL %0
            KILL %1
            %2 = 1
        };"
    );
}

#[test]
fn test_ir_block() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_block",
        req,
        "define test (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            GOTO 1
            BLOCK 1
            %1 = 2
        };"
    );
}

#[test]
fn test_ir_if_else() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_if_else",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = 2
            %2 = 3
            GOTO 3
            BLOCK 2
            %2 = 4
            GOTO 3
            BLOCK 3
            %0 = %2
        };
        "
    );
}

#[test]
fn test_ir_if() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_if",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 0
            %1 = 2
            %2 = 3
        };
        "
    );
}

#[test]
fn test_ir_loop() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_loop",
        req,
        "define main (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            %1 = 2
            GOTO 0 
        };
        "
    );
}

#[test]
fn test_ir_table() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_table",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            %1 = 2
            %2 = 3
            BLOCK 1
            %1 = 2
            %2 = 3
            BLOCK 2
            %2 = 4
            TABLE GOTO 0 1 2 ELSE GOTO 2
        };
        "
    );
}

#[test]
fn test_global_get_and_set() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_get_and_set",
        req,
        "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1) {
        %1 = %-1
        %-2 = %1
        };
    "
    );
}

#[test]
fn test_global_set() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_set",
        req,
        "
        define 0 (param %0) (result 0) (define %-1 %0 %1) {
            %0 = 1
            %-1 = %0
            %1 <- CALL 1()
        };

        define 1 (param) (result 1) (define %-1 %0) {
            %0 = %-1
            RETURN %0;
        };
    "
    );
}

#[test]
fn test_global_get_and_set_multiple_functions() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_get_and_set_multiple_functions",
        req,
        "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1 %2) {
        %1 = %-1
        %-2 = %1
        %2 = 1
        %0 <- CALL 1 (%2)
        %1 <- CALL 2 ()
        };

        define 1 (param %0) (result 1) (define %-2 %0) {
        %0 = %-2
        RETURN %0;
        };

        define 2 (result 1) (define %-2 %0) {
        %0 = 1
        %-2 = 1
        RETURN %0;
        };
    "
    );
}

#[test]
fn test_global_call() {
    let req = Request {
        variable: None,
        function: "1".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_call",
        req,
        "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1) {
        BLOCK 0
        %1 = %0
        %-1 = %1
        RETURN ;
        };
        define 1 (param %0) (result 0) (define %-2 %-1 %0 %1 %2) {
        BLOCK 1
        %1 = 1
        CALL 0(%1)
        %2 = %0
        %-2 = %2
        RETURN ;
        };
    "
    );
}

#[test]
fn test_global_writes() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_globals_writes",
        req,
        "
        define test (result 0) (define %-1 %0 %2) {
            %0 = 1
            %-1 = %0 
            %2 <- CALL mytest()
        };
        define mytest (param) (result 1) (define %-1 %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN %-1;
        };
    "
    );
}

#[test]
fn test_global_check_order() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_globals_check_order",
        req,
        "
        define test (result 0) (define %-2 %-1 %0 %2) {
            %0 = 1
            %-1 = %0 
            %2 <- CALL mytest()
        };
        define mytest (param) (result 1) (define %-2 %-1 %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN %-1;
        };
    "
    );
}

#[test]
fn test_ir_simple_store() {
    env_logger::init();
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_store",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            STORE FROM %0 OFFSET 0 + %0 ALIGN 2 32
            %0 = 1
         };
    "
    );
}

#[test]
fn test_ir_simple_load() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_load",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            %1 = LOAD OFFSET 0 + %0 ALIGN 0
         };
    "
    );
}

#[test]
fn test_memory_store() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_memory_store",
        req,
        "
        define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7 %8 %9) {
        BLOCK 0
        %1 = -12345
        STORE FROM %1 OFFSET 0 + %0 ALIGN 2 32
        %2 = 8
        %3 = -12345
        STORE FROM %3 OFFSET 0 + %2 ALIGN 3 64
        %5 = 8
        %6 = -12345
        STORE FROM %6 OFFSET 0 + %5 ALIGN 2 32
        %7 = 8
        %8 = -12345
        STORE FROM %8 OFFSET 0 + %7 ALIGN 3 64
        RETURN ;
        };
    "
    );
}

#[test]
fn test_memory_load() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_memory_load",
        req,
        "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE FROM %1 OFFSET 0 + %0 ALIGN 2 32
        %4 = 8
        %5 = LOAD OFFSET 0 + %4 ALIGN 0
        %6 = 8
        %7 = LOAD OFFSET 0 + %6 ALIGN 0
        KILL %7
        KILL %6
        RETURN ;
       }; 
    "
    );
}

#[test]
fn test_memory_load_different_functions() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_memory_load_different_functions",
        req,
        "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE FROM %1 OFFSET 0 + %0 ALIGN 2 32
        %2 <- CALL 1 ()
        RETURN ;
       }; 

       define 1 (result 1) (define %0 %1) {
        %1 = 8
        %0 = LOAD OFFSET 0 + %1 ALIGN 0
        RETURN %0;
       };
    "
    );
}

#[test]
fn test_memory_load_different_functions2() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_memory_load_different_functions2",
        req,
        "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE FROM %1 OFFSET 0 + %0 ALIGN 2 32
        %2 <- CALL 1 ()
        STORE FROM %1 OFFSET 1 + %0 ALIGN 2 32
        %3 <- CALL 1 ()
        RETURN ;
       }; 

       define 1 (result 1) (define %0 %1) {
        %1 = 8
        %0 = LOAD OFFSET 0 + %1 ALIGN 0
        RETURN %0;
       };
    "
    );
}

