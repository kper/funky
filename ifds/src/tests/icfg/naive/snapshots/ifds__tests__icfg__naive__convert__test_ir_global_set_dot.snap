---
source: src/tests/icfg/naive/convert.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (13) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (14) at (2, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (15) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (16) at (2, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (17) at (1, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (18) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (19) at (1, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (20) at (2, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (32) at (2, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (33) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (34) at (2, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (35) at (1, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (36) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (39) at (2, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (41) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (42) at (1, 2) { };
\node[font=\tiny] (note_3) at (-1.5, 0.5) {Assign(0, -1)};
\node[font=\tiny] (note_4) at (-1.5, 1.5) {Return([0])};
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (0) at (4, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (1) at (7, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (2) at (5, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (3) at (4, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (4) at (4, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (5) at (6, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (6) at (5, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (7) at (6, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (8) at (4, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (9) at (7, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (10) at (6, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (11) at (4, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (12) at (7, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (21) at (4, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (22) at (7, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (23) at (5, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (24) at (6, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (25) at (4, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (26) at (6, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (27) at (5, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (28) at (7, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (29) at (4, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (30) at (7, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (31) at (6, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (37) at (4, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (38) at (5, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (40) at (7, 3) { };
\node[font=\tiny] (note_0) at (2.5, 0.5) {Const(0, 1.0)};
\node[font=\tiny] (note_1) at (2.5, 1.5) {Assign(-1, 0)};
\node[font=\tiny] (note_2) at (2.5, 2.5) {Call(1, [], [1])};
		\path[->] (0) edge (4);
		\path[->] (1) edge (22);
		\path[->] (2) edge (6);
		\path[->] (0) edge (5);
		\path[->] (4) edge (8);
		\path[->] (5) edge (10);
		\path[->] (6) edge (27);
		\path[->] (5) edge (9);
		\path[->] (8) edge (29);
		\path[->] (9) edge (30);
		\path[->] (10) edge (31);
		\path[->, green] (8) [bend left] edge node { } (13);
		\path[->, green] (9) [bend left] edge node { } (14);
		\path[->] (13) edge (15);
		\path[->] (14) edge (16);
		\path[->] (14) edge (17);
		\path[->] (15) edge (36);
		\path[->] (16) edge (39);
		\path[->] (17) edge (42);
		\path[->, red] (15) [bend right] edge  node { } (29);
		\path[->, red] (17) [bend right] edge  node { } (38);
		\path[->, red] (16) [bend right] edge  node { } (30);


                \end{tikzpicture}
            \end{document}
